src/chainstate/stacks/boot/pox-2.clar

```diff
@@ -327,18 +327,48 @@
                                  cycle-entry-total-ustx) })
             (ok (merge data { cycle: (+ u1 cycle)} )))))
 
+(define-private (get-event-info (stacker principal))
+    (let (
+        (stacker-info (stx-account stacker))
+        (total-balance (stx-get-balance stacker))
+    )
+        {
+            ;; The principal of the stacker 
+            stacker: stacker,
+            ;; The current available balance, as string quoted micro-STX */
+            balance: total-balance,
+            ;; [PoX] The amount of locked STX, as string quoted micro-STX. Zero if no tokens are locked. */
+            locked: (get locked stacker-info),
+            ;; [PoX] The burnchain block height of when the tokens unlock. Zero if no tokens are locked. */
+            burnchain-unlock-height: (get unlock-height stacker-info),
+        }
+    )
+)
+
 ;; This method is called by the Stacks block processor directly in order to handle the contract state mutations
 ;;  associated with an early unlock. This can only be invoked by the block processor: it is private, and no methods
 ;;  from this contract invoke it.
 (define-private (handle-unlock (user principal) (amount-locked uint) (cycle-to-unlock uint))
     (let ((user-stacking-state (unwrap-panic (map-get? stacking-state { stacker: user })))
           (first-cycle-locked (get first-reward-cycle user-stacking-state))
-          (reward-set-indexes (get reward-set-indexes user-stacking-state)))
+          (reward-set-indexes (get reward-set-indexes user-stacking-state))
+          (event-info (get-event-info user)))
+        ;; Check that the expected amount locked is equal to the actual amount locked
+        (asserts! (is-eq amount-locked (get locked event-info)) (err ERR_STACKING_CORRUPTED_STATE))
         ;; iterate over each reward set the user is a member of, and remove them from the sets. only apply to reward sets after cycle-to-unlock.
         (try! (fold fold-unlock-reward-cycle reward-set-indexes (ok { cycle: first-cycle-locked, first-unlocked-cycle: cycle-to-unlock, stacker: user })))
         ;; Now that we've cleaned up all the reward set entries for the user, delete the user's stacking-state
         (map-delete stacking-state { stacker: user })
-        (ok true)))
+
+        (ok (merge event-info {
```

Comment:
Here and elsewhere, it looks like you're returning the _prior_ state of the user?  i.e. `event-info` represents the state of the user's balances and lock-ups at the _start_ of the function call?

If so, then I'm wondering if downstream clients need this information?  IIRC @zone117x said something about this being unnecessary in the blockchain call today.  Maybe instead they can return the _current_ state of the user, after it's been modified?

The reason I'm proposing this is that if we can do this, then we can move this entire return value out of `.pox-2` and instead generate it after the contract-call finishes in the special-case contract call handler.

Reply:
Chiming in here to note that Jude is correct -- downstream clients need the current state rather than the prior state.